File: ios/CamerasApp/Networking/RetryManager.swift
import Foundation

final class RetryManager {
    private var currentDelay: TimeInterval = AppConfig.retryInitialDelay
    private var attempts = 0
    private var rateLimitedUntil: Date?

    var isRateLimited: Bool {
        guard let until = rateLimitedUntil else { return false }
        return Date() < until
    }

    func handleRateLimit(retryAfter: TimeInterval) {
        rateLimitedUntil = Date().addingTimeInterval(retryAfter)
    }

    func handleFailure() -> TimeInterval? {
        attempts += 1
        guard attempts <= AppConfig.retryMaxAttempts else { return nil }
        let delay = currentDelay
        currentDelay = min(currentDelay * 2, AppConfig.retryMaxDelay)
        return delay
    }

    func reset() {
        currentDelay = AppConfig.retryInitialDelay
        attempts = 0
        rateLimitedUntil = nil
    }
}

File: android/app/src/main/java/com/iceblox/app/network/RetryManager.kt
package com.iceblox.app.network

import com.iceblox.app.config.AppConfig

class RetryManager {
    private var attempts = 0
    private var rateLimitedUntil = 0L

    val isRateLimited: Boolean
        get() = System.currentTimeMillis() < rateLimitedUntil

    fun handleFailure(): Long? {
        if (attempts >= AppConfig.RETRY_MAX_ATTEMPTS) return null
        val delay = minOf(
            AppConfig.RETRY_INITIAL_DELAY_MS * (1L shl attempts),
            AppConfig.RETRY_MAX_DELAY_MS
        )
        attempts++
        return delay
    }

    fun handleRateLimit(retryAfterSeconds: Long) {
        rateLimitedUntil = System.currentTimeMillis() + retryAfterSeconds * 1000
    }

    fun reset() {
        attempts = 0
    }
}
